src/main.js

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import * as core from '@actions/core';
+import * as fs from "fs/promises";
+import * as https from "https";
+import * as path from "path";
+import * as child_process from "child_process";
+
+async function getFileTo(url, outFile) {
+    let file = await fs.open(outFile, 'w');
+    let getFn = url => (resolve, reject) => {
+        let req = https.get(url, (res) => {
+            if(res.statusCode == 301 || res.statusCode == 302) {
+                // Follow redirects.
+                getFn(res.headers.location)(resolve, reject)
+                return;
+            }
+            if(res.statusCode !== 200) {
+                reject(new Error(`Unexpected response: ${res.statusCode} at ${url}`));
+            }
+            let datas = [];
+            res.on('data', (data) => {
+                // It seems that it's possible for these to resolve out of order on certain platforms if we just write them here.
+                datas.push(data);
+            });
+            res.on('end', async () => {
+                for(let data of datas) {
+                    await file.write(data);
+                }
+                file.close();
+                resolve();
+            });
+        });
+        req.on('error', reject);
+    };
+    await new Promise(getFn(url));
+    return outFile;
+}
+
+function exec(command) {
+    return new Promise((resolve, reject) => {
+        let proc = child_process.exec(command);
+        let output = "";
+        proc.stderr.pipe(process.stderr);
+        proc.stdout.pipe(process.stdout);
+        proc.stdout.on('data', (chunk) => {
+            output += chunk;
+        });
+        proc.on('exit', () => {
+            resolve({
+                exitCode: proc.exitCode,
+                output: output
+            });
+        });
+        proc.on('error', () => {
+            reject({
+                exitCode: -1,
+                output: output
+            });
+        });
+    });
+}
+
+async function extractFile(file) {
+    // Assumes unzip and tar is installed.
+    let ext = path.extname(file);
+    if(ext === ".zip") {
+        // Might need to use powershell Extract-Archive on Windows?
+        await exec(`unzip ${file}`);
+    }
+    else if(ext === ".gz") {
+        await exec(`tar xvf ${file}`);
+    }
+    else {
+        throw new Error(`Unexpected file extension ${ext}`);
+    }
+    return file;
+}
+
+function validatePlatform(platform) {
+    let platformList = [
+        'linux',
+        'macOS',
+        'windows'
+    ];
+    if(!platformList.includes(platform)) {
+        throw new Error(`Platform ${platform} not in list of supported platforms: ${platformList}`);
+    }
+    return platform;
+}
+
+
+function validateArch(platform, arch) {
+    let archMap = {
+        "linux" : [
+            "oneapi",
```

Comment:
`oneapi` is definitely not an architecture.

What about simplifying the input parameters to reasonable defaults? For example, we could remove `architecture`/`platform` and install ISPC based on host/arch detection:

- windows -> ispc-v1.24.0-windows.zip
- linux -> ispc-v1.24.0-linux.tar.gz
- any macos -> ispc-v1.24.0-macOS.universal.tar.gz
- (when aarch64 linux runners became available) linux aarch64 -> ispc-v1.24.0-linux.aarch64.tar.gz

As for me, this seems sufficient for the main use cases. If and when someone needs macOS-x86, we can add an option to install it.

Reply:
Open to a rename for what this field refers to in the download path. 

After adding auto-detection, I'd prefer to still have a method to specify the platform, otherwise there is no to test locally on Mac without Mac hardware.

Reply:
` require('node:os');`

`os.arch` and `os.platform` look sufficient to do detection here